addons/sale/models/sale.py

```diff
@@ -202,6 +202,13 @@ def _get_payment_type(self):
         ('confirmation_date_conditional_required', "CHECK( (state IN ('sale', 'done') AND confirmation_date IS NOT NULL) OR state NOT IN ('sale', 'done') )", "A confirmed sales order requires a confirmation date."),
     ]
 
+    @api.constrains('company_id', 'order_line')
+    def _check_company_id(self):
+        for order in self:
+            for line in order.order_line:
```

Comment:
```
for product_id in order.mapped('product_id'):
    if product_id.company_id and product_id.company_id != order.company_id:
        raise"blabla"
```

Reply:
Can we make this more readable ? (Instead of long text)
![screenshot-551261-34342-0b6909 runbot22 odoo com-2019 06 28-23-06-16](https://user-images.githubusercontent.com/16624719/60360441-5ac0c500-99f9-11e9-864d-feee83a7e8fc.jpg)

Reply:
```python
for order in self:
  companies = order.mapped('order_line.product_id.company_id')
  if companies and companies != self.company_id:
    raise ...
```
instead :D

Reply:
> self.company_id


Chances of `expected singleton`

Reply:
(Not tried this one yet)

    for order in self:
        if order.mapped('order_line.product_id.company_id').filtered(lambda company_id: company_id != order.company_id):
            raise ...


But not that much readable in IMO.
I will let you decide the final code. :+1:

Reply:
Indeed, I missed changing a `self` to `order` :D (and edited it).

Your code sample doesn't consider the case where the product has no company (and is available inter company) while the order company_id is set, which will raise an Error when not needed (because False is different than a set id) ;).  
And why filtering the companies ? raising `"Your order SO123 belongs to company 1 and contains products from companies 1,2,3"` is clear enough for me.

Reply:
@sswapnesh my bad, read your code too fast.  I wanted to say that you could compare the recordsets instead of using a filtered, but your solution also works perfectly because it also supports the case where a product has no company :).

Reply:
>  read your code too fast.

Can understand, This happens sometimes to me also.

> I wanted to say that you could compare the recordsets instead of using a filtered

I used `filtered` so skip iteration over companies (and checking `if company_id != order_id.company_id`).